sdk/src/main/java/pt/ua/dicoogle/sdk/IndexerInterface.java

```diff
@@ -67,8 +93,63 @@ public default boolean handles(URI path) {
     /**
      * Removes the indexed file at the given path from the database.
      * 
+     * Unlike the other indexing tasks,
+     * this operation is synchronous
+     * and will only return when the operation is done.
+     * 
      * @param path the URI of the document
      * @return whether it was successfully deleted from the database
      */
     public boolean unindex(URI path);
+
+    /**
+     * Removes indexed files from the database in bulk.
+     *
+     * The default implementation unindexes each item one by one
+     * in a non-specified order via {@linkplain #unindex(URI)},
+     * but indexers may implement this as
+     * one or more individual operations in batch,
+     * thus becoming faster than unindexing each item individually.
+     * 
+     * Like {@linkplain index},
+     * this operation is asynchronous.
+     * One can keep track of the unindexing task's progress
+     * by passing a callback function as the second parameter.
+     *
+     * @param uris the URIs of the items to unindex
+     * @param progressCallback an optional function (can be `null`),
+     *        called for every batch of items successfully unindexed
+     *        to indicate early progress
+     *        and inform consumers that
+     *        it is safe to remove or exclude the unindexed item
+     * @return an asynchronous task object returning
+     *         a report containing which files were not unindexed,
+     *         and whether some of them were not found in the database
+     * @throws IOException if an error occurred
+     *         before the unindexing operation could start,
+     *         such as when failing to access or open the database
+     */
+    public default Task<UnindexReport> unindex(Collection<URI> uris, Consumer<Collection<URI>> progressCallback)
+            throws IOException {
+        Objects.requireNonNull(uris);
+        return new Task<>(() -> {
+            List<FailedUnindex> failures = new ArrayList<>();
+            for (URI uri : uris) {
+                try {
+                    if (unindex(uri)) {
+                        // unindexed successfully
+                        if (progressCallback != null) {
+                            progressCallback.accept(Collections.singleton(uri));
+                        }
+                    } else {
+                        // failed to unindex, reason unknown
+                        failures.add(new FailedUnindex(Collections.singleton(uri), null));
+                    }
+                } catch (Exception ex) {
+                    failures.add(new FailedUnindex(Collections.singleton(uri), ex));
+                }
+            }
+            return UnindexReport.withFailures(failures);
```

Comment:
If `progressCallback` is null, the UnindexServlet will log “Finished unindexing 0 out of `N` files” for any bulk index, even if there were successful operations.

Is this intended behavior?

Reply:
Well, not really, but I'm not sure if this is a problem with the core components or with the plugins. I updated `UnindexReport` so that the file count is well calculated, but I feel that that log line should not depend on `progressCallback`.